include/multiverso/util/quantization_util.h

```diff
@@ -4,137 +4,162 @@
-            auto& blob = blobs[i + 1];
-            outputs->push_back(is_compressed ?
-                std::move(de_compress(blob, size)) : blob);
+      std::vector<Blob>* outputs) override {
+      CHECK_NOTNULL(outputs);
+      // the blobs at least contain row_index blob
+      CHECK(blobs.size() > 1);
+      outputs->clear();
+      outputs->push_back(blobs[0]);
+      size_t data_blobs_size = skip_option_blob_? blobs.size() - 1:blobs.size();
+      
+      if (data_blobs_size > 1){
+        auto& size_blob = blobs[1];
+        for (auto i = 2; i < data_blobs_size; i++) {
+          // size info (compressed ? size : -1)
+          auto is_compressed = size_blob.As<index_type>(i - 2) >= 0;
+          auto size = is_compressed ?
+            size_blob.As<index_type>(i - 2) : blobs[i].size();
+          auto& blob = blobs[i];
+          outputs->push_back(is_compressed ?
+            std::move(DeCompress(blob, size)) : blob);
         }
+      }
+      if (skip_option_blob_){
+        //adding the original option to the output
+        outputs->push_back(blobs[blobs.size() - 1]);
+      }
     }
 
- protected:
-    bool try_compress(const Blob& in_blob,
-        Blob* out_blob) {
-        CHECK_NOTNULL(out_blob);
-        CHECK(sizeof(data_type) == sizeof(index_type));
-        auto data_count = in_blob.size<data_type>();
-        auto non_zero_count = 0;
-        for (auto i = 0; i < data_count; ++i) {
-            if (std::abs(in_blob.As<data_type>(i)) > clip_value) {
-                ++non_zero_count;
-            }
-        }
-
-        if (non_zero_count * 2 >= data_count)
-            return false;
+  protected:
 
-        if (non_zero_count == 0) {
-            // Blob does not support empty content,
-            //  fill the blob with first value
-
-            Blob result(2 * sizeof(data_type));
+    bool TryCompress(const Blob& in_blob,
+      Blob* out_blob) {
+      CHECK_NOTNULL(out_blob);
+      auto data_count = in_blob.size<data_type>();
+      auto non_zero_count = 0;
+      for (auto i = 0; i < data_count; ++i) {
+        if (std::abs(in_blob.As<data_type>(i)) > clip_value_) {
+          ++non_zero_count;
+        }
+      }
+
+      if (non_zero_count * 2 >= data_count)
+        return false;
+
+      if (non_zero_count == 0) {
+        // Blob does not support empty content,
+        //  fill the blob with first value
+
+        Blob result(2 * sizeof(data_type));
+        // set index
+        result.As<index_type>(0) = 0;
+        // set value
+        result.As<data_type>(1) = in_blob.As<data_type>(0);
+        *out_blob = result;
+      }
+      else {
+        Blob result(non_zero_count * 2 * sizeof(data_type));
+        auto result_index = 0;
+        for (auto i = 0; i < data_count; ++i) {
+          auto abs_value = std::abs(in_blob.As<data_type>(i));
+          if (abs_value > clip_value_) {
             // set index
-            result.As<index_type>(0) = 0;
+            result.As<index_type>(result_index++) = i;
```

Comment:
it you want to traverse blob, the better choice is cast to raw pointer and than access it. Using As method in a for loop will result in performance overhead since it will cast every time.